File: new-21-game/solution.py
class Solution:
    def new21Game(self, n: int, k: int, maxPts: int) -> float:
        if k==0 or k+maxPts<=n:
            return 1.0
        window=0
        for i in range(k,k+maxPts):
            window+=1 if i<=n else 0
        dp={}
        for i in range(k-1,-1,-1):
            dp[i]=window/maxPts
            rem=0
            if i+maxPts<=n:
                if i+maxPts not in dp:
                    dp[i+maxPts]=1
                rem=dp[i+maxPts]
            window+=dp[i]-rem
        return dp[0]

File: new-21-game/solution.cpp
class Solution {
public:
    double new21Game(int n, int k, int maxPts) {
        if (k == 0 || k + maxPts <= n) {
            return 1.0;
        }
        double window = 0.0;
        for (int i = k; i < k + maxPts; ++i) {
            window += (i <= n) ? 1 : 0;
        }
        vector<double> dp(k + maxPts + 1, 0);
        for (int i = k - 1; i >= 0; --i) {
            dp[i] = window / maxPts;
            double rem = 0.0;
            if (i + maxPts <= n) {
                if (dp[i + maxPts] == 0) {
                    dp[i + maxPts] = 1;
                }
                rem = dp[i + maxPts];
            }
            window = window + dp[i] - rem;
        }
        return dp[0];
    }
};